Add `only_return_final=True` for improved efficiency

joelmoniz opened the pull request:
The LSTM example in text generation seems to use only the final sequential output. This PR sets the `only_return_final` parameter to `True`, which offers improved efficiency if intermediate outputs aren't required.

f0k commented:
This looks wrong. Did you compare the results? If the first layer already returns the final state only, there's nothing the second layer can scan over. And the SliceLayer should be omitted.

ebenolson commented:
Yeah I haven't tested this, but I think a better change would be to get rid of the SliceLayer - using the loss from every timestep will change things a bit, but it will train much faster.

joelmoniz commented:
Thanks for the review! Extremely sorry about that, my bad. I have made the following changes:

1. The first layer should not return only the final state. I have updated my first commit to reflect this. Not sure how I missed that, sorry.
2. Indeed, the `SliceLayer` is no longer required. I have added in a commit making this change. I have also only commented out the `SliceLayer`, and explained why it is not required. Could you please let me know if I should remove these lines completely instead?

I have gone on to time the earlier implementation and the slightly modified one this PR proposes. 
There does seem to be some performance improvement, but it is extremely small- I observed an average speedup of around 0.25%.

f0k commented on examples/lstm_text_generation.py:
```diff
@@ -127,16 +127,16 @@ def main(num_epochs=NUM_EPOCHS):
-    # Since we are only interested in the final prediction, we isolate that quantity and feed it to the next layer. 
-    # The output of the sliced layer will then be of size (batch_size, N_HIDDEN)
-    l_forward_slice = lasagne.layers.SliceLayer(l_forward_2, -1, 1)
+    # A SliceLayer is no longer required, since l_forward layer creates an output of dimension (batch_size, N_HIDDEN) because
```
If someone reads this code, they will wonder what `no longer` refers to. Also the next comment starts with `The sliced output`, which does not fit the code any more. Can you amend your commit to just drop the `SliceLayer` code and comment, and adapt the next comment?

joelmoniz replied:
Ah, I had simply commented out the `SliceLayer` (it is present in the comments directly below this diff block), but I agree that it can get a little confusing. I've modified all the comments accordingly.

ebenolson commented:
I'm not too familiar with the RNN layers, but I think the performance makes sense - all the intermediate values still need to be calculated even if they aren't returned. The docs refer to a memory usage benefit from `only_return_final`, not performance.

The optimization I was referring to would be to use the outputs from every step in calculating the loss, which should reduce the number of iterations for convergence roughly by a factor `SEQ_LENGTH`. However, it's a bit different since the earlier outputs can't be backpropagated as far. Also it would probably be necessary to carry the hidden states from batch to batch. There's an example of doing it this way [here](https://github.com/ebenolson/pydata2015/blob/master/4%20-%20Recurrent%20Networks/RNN%20Character%20Model%20-%202%20Layer.ipynb), but changing all that would be a bit much for this PR.

If you can go ahead and make the comment changes, I think this is good - even if it's not faster, using `only_return_final` makes it a bit clearer.

joelmoniz commented:
@ebenolson Ah, indeed. I had automatically assumed that you were referring to optimization in terms of time and not memory with respect to `only_return_final`. 

I also now understand the optimization you were referring to with respect to improving convergence. I got a little confused, and assumed that by "train faster" you meant less time per epoch and not less number of epochs overall due to faster convergence. My bad!

f0k commented:
LGTM, but the two commits need to be squashed. The first commit alone doesn't do anything useful, because the slice layer will slice out a single hidden unit then.

joelmoniz commented:
Ah yes, I agree. I've squashed them up now.

f0k commented:
Oh, seems this one was forgotten -- everything looks good, merging!
